docs/site/REST-middleware-sequence.md

```diff
@@ -0,0 +1,846 @@
+
+  constructor(
+    @inject(SequenceActions.FIND_ROUTE) protected findRoute: FindRoute,
+    @inject(SequenceActions.PARSE_PARAMS) protected parseParams: ParseParams,
+    @inject(SequenceActions.INVOKE_METHOD) protected invoke: InvokeMethod,
+    @inject(SequenceActions.SEND) public send: Send,
+    @inject(SequenceActions.REJECT) public reject: Reject,
+  ) {}
+
+  async handle(context: RequestContext) {
+    try {
+      const {request, response} = context;
+      const finished = await this.invokeMiddleware(context);
+      if (finished) return;
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+      this.send(response, result);
+    } catch (err) {
+      this.reject(context, err);
+    }
+  }
+}
+```
+
+The legacy `Sequence` is now deprecated but it should continue to work without
+any changes.
+
+If you have never customized `src/sequence.ts` generated by `lb4 app`, or you
+have only modified the code by adding `authenticate` action per our docs, you
+can simply migrate to the middleware based sequence by replacing the content of
+`src/sequence.ts` with the code below.
+
+```ts
+import {MiddlewareSequence} from '@loopback/rest';
+
+export class MySequence extends MiddlewareSequence {}
+```
+
+If you have other actions in your sequence, you'll have to write a middleware to
+wrap your action and register it with the middleware chain for the
+middleware-based sequence. For example, the `invokeMethod` action is a function
+with the following signature:
+
+```ts
+export type InvokeMethod = (
+  route: RouteEntry,
+  args: OperationArgs,
+) => Promise<OperationRetval>;
+```
+
+The middleware will look like:
+
+```ts
+async (ctx, next) => {
+  // Locate or inject input parameters from the request context
+  const route: RouteEntry = await ctx.get(RestBindings.Operation.ROUTE);
+  const params: OperationArgs = await ctx.get(RestBindings.Operation.PARAMS);
+  const retVal = await this.invokeMethod(route, params);
+  // Bind the return value to the request context
+  ctx.bind(RestBindings.Operation.RETURN_VALUE).to(retVal);
+  return retVal;
+};
+```
+
+## Extend the middleware sequence
+
+### Add middleware to the chain
+
+LoopBack middleware (including Express middleware) can be added to the sequence.
+See
+[Middleware](Middleware.md#register-middleware-to-be-executed-by-invokemiddleware-actions)
+for more details.
+
+### Sort middleware by groups
+
+The middleware for the sequence are executed in the order of groups in a
+cascading style. The order of groups is determined by two factors:
+
+1.  The relative order specified for a middleware binding.
+
+    - upstreamGroups: An array of group names that should be upstream to this
+      middleware
+
+      ```ts
+      @bind(
+        asMiddleware({
+          chain: RestTags.REST_MIDDLEWARE_CHAIN,
+          group: 'authentication',
+          upstreamGroups: ['cors', 'findRoute'],
```

Comment:
A question: would `asMiddleware` make sure the group name is really after the upstream groups? same question for downstream groups.

Reply:
`asMiddleware` adds tags to the binding to express its ordering requirement and we will use these partial rules to line up all middleware.